### DISCUSSION TITLE
Scalability [graph usability] - Usability of the UI graph

### DISCUSSION THREAD

**Discussion by lucasponce** [kiali/kiali#4518] [Roadmap]
Related to https://github.com/kiali/kiali/discussions/4080

We have started several scalability discussions which good topics but I think those are more located in the backend side and I'd like to share some thoughts about the frontend and the user experience.

I think that the current topology graph works well in namespaces with few services but it starts to feel some "UX" pressure when we try to visualize several namespaces together or just one/two namespaces but more services than the usual.

In particular I guess there are several aspects that we could discuss, I don't have specific proposals but I'd try to enumerate some "pain" points from my user experience:

- No need to think on dozens of namespaces, there are usability challenges when using two namespaces (i.e. try the demo error-rates app and try to visualize alpha and beta at same time).
- Layout should try to be more intelligent and somehow try to fit better, the grouping boxes perhaps could be improved to optimize better the space.
- Graph should detect when something is "readable" and it shouldn't show elements that user can't read (i.e. if I can't read a label, just hide it, but don't render a minuscule label with tiny font, that forces the user to zoom in/out a lot of times creating some visual fatigue).
- The "hover" effect to highlight elements create some blink effect, personally the pattern to highlight an element (node or edge) is good, but the current implementation creates some visual fatigue, perhaps it could be researched other alternative to get same result.
- Perhaps the sidepanel could benefit of other font changes, to help the user to read better the content, or to change the current design and allow a float panel closer to the node when click.
- I share impression that we have "too many" clicks that perhaps those can be researched (i.e. consolidate the node graph in the details page ?).
- Also, there are "hidden" events that are difficult for the user, perhaps we could introduce some decorators to indicate that from a node you could click here or there to perform some operation or navigate into details.
- The Display options is growing and I think as a user I'd prefer to have more menus with less options rather than a single big list.
- Tooltip information is useful but hard to read (too many lines for a tooltip, some descriptions are excellent for a help page, but I have impression that the tooltip is not the perfect format to display that info).

These "pain points" are not necessarily all related to "big" deployments, I think when we have a namespace with 10-20 services those are somehow visible.

I don't have specific suggestions, but I think that the current discussions were missing a specific frontend thread.
I'll try to add / record some specific use cases / examples when I have a demo available.

**Comment by jshaughn** [comment]
Several good observations here, comments on a few of them:

> Layout should try to be more intelligent and somehow try to fit better, the grouping boxes perhaps could be improved to optimize better the space.

This would be hard, I think. It would mean trying to optimize an existing layout or create a new one.

> Graph should detect when something is "readable" and it shouldn't show elements that user can't read (i.e. if I can't read a label, just hide it, but don't render a minuscule label with tiny font, that forces the user to zoom in/out a lot of times creating some visual fatigue).

I've seen this suggested multiple times, and I also think we could do something in this area.  I'll create an issue and play with this idea.  This could possibly help with the layout issues above for larger graphs.  (https://github.com/kiali/kiali/issues/4521)

> The "hover" effect to highlight elements create some blink effect, personally the pattern to highlight an element (node or edge) is good, but the current implementation creates some visual fatigue, perhaps it could be researched other alternative to get same result.

I agree the flashing can be annoying.  This has been an area of refinement and changes for a long time and I still think it could be better. 
 A few options to consider:

- remove the hover highlighting completely and just defer to use selecting (via click)
- reduce the hover highlighting to just the immediate node or edge, not the "neighborhood". (This would be an easy change, perhaps I'll try it locally.)
- remove highlighting on the graph but instead dedicate a fixed area (maybe the top of the full-graph-side-panel) for highlighting info about the hovered-element. This would mainly just be an easy way to read a label.
- try and introduce a hover "delay" to only highlight if the user spends some time hovering.   The problem here is that I'm not aware of a built-in mechanism for adding a delay to the `mouseover` event, so  we'd likely have to start our own timer.

> Perhaps the sidepanel could benefit of other font changes, to help the user to read better the content, or to change the current design and allow a float panel closer to the node when click.

Not too long ago we did a complete overhaul of the side-panel, and it is much more consistent and readable than in the past.  For me, it's OK and a floating panel will occlude the graph.  TBH, I'd focus elsewhere.
